src/api/libcellml/parser.h

```diff
@@ -0,0 +1,168 @@
+    ModelPtr parseModel(const std::string &input);
+
+    /**
+     * @brief Updates an existing model with the attributes from a @c std::string.
+     *
+     * Updates the argument @p model with CellML entities and attributes
+     * from the argument @p input @c std::string. Any entities, objects,
+     * or attributes of with matching names in the model will be
+     * overwritten by those loaded from the @p input @c std::string.
+     *
+     * @param model The @c ModelPtr to update.
+     * @param input The string to parse into the @p model.
+     */
+    void updateModel(const ModelPtr &model, const std::string &input);
+
+private:
+    /**
+     * @brief Load the @p model with attributes parsed from the XML @p node.
+     *
+     * Loads the argument @p model with attributes and objects parsed from
+     * the argument XML @p node. Existing objects/attributes with matching
+     * names in the @p model will be overwritten by those from the XML @p node.
+     *
+     * @param model The @c ModelPtr to update.
+     * @param node The @c XmlNodePtr to load the @p model from.
+     */
+    void loadModel(const ModelPtr &model, const XmlNodePtr &node);
+
+    /**
+     * @brief Load the @p component with attributes parsed from the XML @p node.
+     *
+     * Loads the argument @p component with attributes and objects parsed from
+     * the argument XML @p node. Existing objects/attributes with matching
+     * names in the @p component will be overwritten by those from the XML @p node.
+     *
+     * @param component The @c ComponentPtr to update.
+     * @param node The @c XmlNodePtr to load the @p component from.
+     */
+    void loadComponent(const ComponentPtr &component, const XmlNodePtr &node);
+
+    /**
+     * @brief Load the @p model with attributes parsed from the connection XML @p node.
+     *
+     * Loads the argument @p model with CellML connection information parsed from
+     * the argument XML @p node. If any variable equivalence relationships already
+     * exist in the @p model, the connection information parsed from the XML @p node
+     * will be added to this network.
+     *
+     * @param model The @c ModelPtr to update.
+     * @param node The @c XmlNodePtr to load the connection from.
+     */
+    void loadConnection(const ModelPtr &model, const XmlNodePtr &node);
+
+    /**
+     * @brief Load the @p model with attributes parsed from the encapsulated XML node.
+     *
+     * Loads the argument @p model with CellML encapsulation information parsed from
+     * the argument XML @p parentComponentNode. If any encapsulation relationships already
+     * exist in the @p model, the encapsulation information parsed from the XML @p node
+     * will be added to this network.
+     *
+     * @param model The @c ModelPtr to update.
+     * @param parentComponentNode The parent @c XmlNodePtr of the encapsulation relationship.
+     */
+    void loadEncapsulation(const ModelPtr &model, XmlNodePtr &parentComponentnode);
```

Comment:
Yes, I did see that `parentComponentnode = parentComponentnode->getNext()`, but couldn't we still have:

```
void loadEncapsulation(const ModelPtr &model, const XmlNodePtr &node);
```

for consistency (with the rest of the `loadXXX()` methods) and then, within the implementation of `loadEncapsulation()`, have something like:

```
XmlNodePtr parentComponentnode = node;
...
parentComponentnode = parentComponentnode->getNext();
```

It would be nice if that was to be possible. (I like things to be consistent and `loadEncapsulation()`'s signature just isn't consistent with that of the other `loadXXX()` methods.)